context/app/static/js/stores/useProvenanceStore.js

```diff
@@ -0,0 +1,44 @@
+/* eslint-disable no-param-reassign */
+
+import create from 'zustand';
+
+import immer from './immerMiddleware';
+
+const useProvenanceStore = create(
+  immer((set, get) => ({
+    steps: [],
+    setSteps: (steps) =>
+      set((state) => {
+        state.steps = steps;
+      }),
+    addSteps: (steps) =>
+      set((state) => {
+        state.steps = [...state.steps, ...steps];
+      }),
+    stitchEntityDescendantSteps: (descendantSteps) => {
+      // check new steps inputs to see if they exist in current steps
+      descendantSteps.forEach((descendantStep) => {
+        descendantStep.inputs.forEach((input) => {
+          get().steps.forEach((step, stepIndex) => {
+            const outputIndex = step.outputs.findIndex((output) => output.name === input.name);
+            // if input exists as an output of an existing step add new step as target
+            if (outputIndex >= 0) {
+              const output = { step: descendantStep.name, name: input.name };
+              set((state) => {
+                state.steps[stepIndex].outputs[outputIndex].target.push(output);
+              });
```

Comment:
I'm thinking I could first loop through the steps outside of this and create a map to get indexes.